static/js/chart/draw.ts

```diff
@@ -724,6 +737,76 @@ function drawLineChart(
         dots.on("click", handleDotClick);
       }
     }
+
+    if (highlightOnHover) {
+      highlight.style("opacity", "0");
+      const highlightedDot = highlight
+        .append("circle")
+        .attr("r", 5)
+        .style("fill", colorFn(dataGroup.label))
+        .style("stroke", "#fff");
+      const highlightedLine = highlight
+        .append("line")
+        .attr("x1", 0)
+        .attr("y1", 0)
+        .attr("x2", 0)
+        .attr("y2", height - MARGIN.bottom)
+        .style("stroke", colorFn(dataGroup.label))
+        .style("stroke-opacity", "0.5");
+      const tooltip = tooltipId
+        ? d3.select(`#${id}`).select(`#${tooltipId}`)
+        : null;
+      svg
+        .on("mouseover", () => {
+          highlight.style("opacity", "1");
+          if (tooltip) {
+            tooltip.style("display", "block");
+          }
+        })
+        .on("mouseout", () => {
+          highlight.style("opacity", "0");
+          if (tooltip) {
+            tooltip.style("display", "none");
+          }
+        })
+        .on("mousemove", () => {
+          // calculation of data point corresponding to mouse position from https://bl.ocks.org/Qizly/8f6ba236b79d9bb03a80.
+          const mouseX = d3.mouse(d3.select("#" + id).node() as HTMLElement)[0];
+          const highlightedTime = xScale.invert(mouseX).getTime();
+          const bisect = d3.bisector((d) => d[0]);
+          let idx = bisect.right(dataset, highlightedTime);
+          if (idx > 0 && idx < dataset.length) {
+            idx =
+              dataset[idx][0] - highlightedTime >
+              highlightedTime - dataset[idx - 1][0]
+                ? idx - 1
+                : idx;
+          } else if (idx === dataset.length) {
+            idx = idx - 1;
+          }
+          if (dataset[idx]) {
+            const datapointX = xScale(dataset[idx][0]);
+            const datapointY = yScale(dataset[idx][1]);
+            highlightedDot.attr(
+              "transform",
+              "translate(" + datapointX + "," + datapointY + ")"
+            );
+            highlightedLine.attr(
+              "transform",
+              "translate(" + datapointX + "," + "0)"
+            );
+            if (handleDotHighlight) {
+              handleDotHighlight(
+                dataGroup.value[idx],
```

Comment:
Do you think it possible to let handleDotHighlight() compute some of these inputs instead of the caller?

Ideally, we can use this function in different chart (e.g. scatter plot) and the call function only need minimal computation.

Reply:
currently, handleDotHighlight is a function that gets passed in from the observationChart component and it's the things that we want to do once a dot has been highlighted. I'm not sure it makes the most sense to compute some of these inputs there because the computed inputs are also needed for the positioning of the line and dot that highlights the point of interest. Maybe I could extract the computation into its own function.

Reply:
This current implementation is specific to line chart and a lot of the code for highlighting is inside the drawLineChart function. Maybe I can see if I can extract things out so that it can be adapted in the future for other charts.

Reply:
I see. It would be nice if this function is used as a shared library and try to decouple it from the drawLineChart. 

But if you feel they are very tightly coupled, it's fine to absorb the function just inside drawLineChart() too.

Reply:
ok, I think I will keep it absorbed inside drawLineChart() for now, but in the future extract it out to be used by multiple-line charts as well